### PR TITLE
Build rust wasm data with Bazel

### PR DESCRIPTION
**Pull request by Shikugawa**
For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Build rust wasm data with bazel. Note that we can't execute this with RBE, and we use private fork of rules_rust because we can't overcome the build problem (see https://github.com/bazelbuild/rules_rust/pull/312) with current version of it.
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by jplevyak** [comment]
Looks good.  Just need to fix the formatting.

File: test/extensions/common/wasm/wasm_vm_test.cc [hunk 1]
```diff
@@ -258,14 +257,14 @@ TEST_P(WasmVmTest, V8Memory) {
...
-      "{{ test_rundir }}/test/extensions/common/wasm/test_data/test_rust.wasm"));
+      "{{ test_rundir }}/test/extensions/common/wasm/test_data/wee8_test_rust.wasm"));
   EXPECT_TRUE(wasm_vm->load(code, GetParam()));
 
   wasm_vm->registerCallback("env", "pong", &pong, CONVERT_FUNCTION_WORD_TO_UINT32(pong));
   wasm_vm->registerCallback("env", "random", &random, CONVERT_FUNCTION_WORD_TO_UINT32(random));
   wasm_vm->link("test");
 
-  EXPECT_EQ(wasm_vm->getMemorySize(), 65536 /* stack size requested at the build-time */);
+  EXPECT_EQ(wasm_vm->getMemorySize(), 1114112 /* stack size requested at the build-time */);
```
Comment (Shikugawa) [inline test/extensions/common/wasm/wasm_vm_test.cc:267]: I don't know to configure stack-size on LLVM via rules_rust. So I changed to follow.

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -364,4 +364,19 @@ REPOSITORY_LOCATIONS = dict(
         strip_prefix = "emsdk-1.39.6",
         urls = ["https://github.com/emscripten-core/emsdk/archive/1.39.6.tar.gz"],
     ),
+    io_bazel_rules_rust = dict(
+        sha256 = "e133b51af2d7d77973f5ebfd7ebc9e347f0e44b32beedd7fe36b77b8fd92a609",
+        strip_prefix = "rules_rust-master-patched",
+        urls = ["https://github.com/Shikugawa/rules_rust/archive/master-patched.zip"],
```
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply bazel/repository_locations.bzl:371]: Since this was merged upstream, could you update this PR?

[the remark this answers is not in the document]
Comment (PiotrSikora) [reply bazel/repository_locations.bzl:371]: Friendly ping.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 3]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
 #[no_mangle]
 pub fn proxy_abi_version_0_1_0() {}
 
-use log::{debug, error, info, trace, warn};
+use log::*;
 
 extern "C" {
-    fn proxy_done() -> u32;
+  fn proxy_done() -> u32;
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:12]: Nit: Unnecessary (and wrong) style change.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 4]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
...
 
 #[no_mangle]
-pub fn proxy_on_context_create(_context_id: u32, _parent_context_id_: u32) {}
-
-#[no_mangle]
-pub fn proxy_on_configure(_context_id: u32, _plugin_configuration_size: u32) -> u32 {
-    trace!("test trace logging");
-    debug!("test debug logging");
-    error!("test error logging");
+pub fn proxy_on_configure(_context_id : u32, _plugin_configuration_size: u32) -> u32 {
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:16]: Nit: Unnecessary (and wrong) style change.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 5]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
...
     warn!("warn configure-test");
     1
 }
 
 #[no_mangle]
-pub fn proxy_on_vm_start(_context_id: u32, _vm_configuration_size: u32) -> u32 {
+pub fn proxy_on_vm_start(_context_id : u32, _vm_configuration_size : u32) -> u32 {
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:22]: Nit: Unnecessary (and wrong) style change.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 6]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
...
+pub fn proxy_on_tick(_context_id : u32) {
     info!("test tick logging");
     unsafe {
         proxy_done();
     }
 }
 
 #[no_mangle]
-pub fn proxy_on_done(_context_id: u32) -> u32 {
+pub fn proxy_on_done(_context_id : u32) -> u32 {
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:38]: Nit: Unnecessary (and wrong) style change.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 7]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
...
     info!("onDone logging");
     0
 }
 
 #[no_mangle]
-pub fn proxy_on_delete(_context_id: u32) {
+pub fn proxy_on_delete(_context_id : u32) {
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:44]: Nit: Unnecessary (and wrong) style change.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 8]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
...
     info!("onDelete logging");
 }
+
+// This function is to compile WebAssembly code with `rust_binary` with Bazel.
+// Envoy wasm runtime won't take care of this type of function.
+fn main() {}
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:50]: Sorry, I don't understand why do we need this? Doesn't `wasm32` target take care of it?

Also, no new line at end of file.

File: test/extensions/common/wasm/wasm_vm_test.cc (hunk 1 above)
Comment (PiotrSikora) [reply test/extensions/common/wasm/wasm_vm_test.cc:257]: I think you could pass it to `rustc` via `rustc_flags` using `-C link-arg=-zstack-size=32768`.

File: test/extensions/bootstrap/wasm/test_data/BUILD [hunk 9]
```diff
@@ -3,15 +3,30 @@ load(
...
 licenses(["notice"])  # Apache 2
 
 envoy_package()
 
-filegroup(
-    name = "modules",
-    srcs = glob(["*.wasm"]),
+rust_library(
+    name = "proxy_wasm",
+    srcs = ["proxy_wasm.rs"],
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/BUILD:15]: I don't have a better place to comment on this, but this `proxy_wasm.rs` is NOT Proxy-Wasm.

You can copy the contents of that file into `logging_rust.rs`, since it implements the ABI directly... or at the very least use a different name to avoid confusion.

Ideally, we should rewrite this test to use `proxy-wasm-rust-sdk`, but that's probably too big change, and should be done in a separate PR.

File: bazel/wasm/wasm.bzl [hunk 10]
```diff
@@ -13,8 +14,13 @@ def _wasm_transition_impl(settings, attr):
         "//command_line_option:collect_code_coverage": "false",
     }
 
-wasm_transition = transition(
-    implementation = _wasm_transition_impl,
+def _wasm_rust_transition_impl(settings, attr):
+    return {
+        "//command_line_option:platforms": "@io_bazel_rules_rust//rust/platform:wasm",
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply bazel/wasm/wasm.bzl:19]: rules_rust doesn't have `wasm32` platform. I think that we need to modify that.

File: test/extensions/bootstrap/wasm/test_data/BUILD (hunk 9 above)
Comment (Shikugawa) [reply test/extensions/bootstrap/wasm/test_data/BUILD:15]: I will add TODO here and fix when proxy-wasm-rust-sdk bazel support is merged.

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 11]
```diff
@@ -6,44 +6,45 @@ use proxy_wasm;
 #[no_mangle]
 pub fn proxy_abi_version_0_1_0() {}
 
-use log::{debug, error, info, trace, warn};
+use log::*;
```
[the remark this answers is not in the document]
Comment (Shikugawa) [reply test/extensions/bootstrap/wasm/test_data/logging_rust.rs:9]: Maybe, `log` crate has private helper macros. In Rust 2018, we can available external macros with `use` syntax. But it works only there is no private helper macros. We should import them when we use log macros. But it will cause more complexity and confusion. In this case, we don't have any problem to use `*`. https://doc.rust-lang.org/edition-guide/rust-2018/macros/macro-changes.html#local-helper-macros

File: test/extensions/bootstrap/wasm/wasm_test.cc [hunk 12]
```diff
@@ -112,14 +112,15 @@ class WasmTestMatrix : public WasmTestBase,
...
+// TODO(shikugawa): Add tests for Rust logging with proxy-wasm-rust-sdk.
 INSTANTIATE_TEST_SUITE_P(RuntimesAndLanguages, WasmTestMatrix,
                          testing::Combine(testing::Values("v8"
 #if defined(ENVOY_WASM_WAVM)
                                                           ,
                                                           "wavm"
 #endif
                                                           ),
-                                          testing::Values("cpp", "rust")));
+                                          testing::Values("cpp")));
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/wasm_test.cc:123]: If this PR disables `wasm_vm_test` and removes `rust` from `wasm_test`, then there are no Rust tests executed at all...

File: test/extensions/common/wasm/BUILD [hunk 13]
```diff
@@ -8,21 +8,26 @@ licenses(["notice"])  # Apache 2
...
+#     srcs = ["wasm_vm_test.cc"],
+#     data = [
+#         "//test/extensions/common/wasm/test_data:test_rust.wasm",
+#     ],
+#     tags = [
+#         # wasm (wee v8 etc) will not compile on Windows
+#         "skip_on_windows",
+#         # Precompiled WASM code which is required to compile with wee8_compiler
+#         # which is needed to build as position independent code. It is not supported on macos.
+#         "skip_on_macos",
```
[the remark this answers is not in the document]
Comment (PiotrSikora) [reply test/extensions/common/wasm/BUILD:22]: Also, you could just skip the precompile step on macOS.

File: test/extensions/bootstrap/wasm/wasm_test.cc (hunk 12 above)
Comment (Shikugawa) [reply test/extensions/bootstrap/wasm/wasm_test.cc:123]: `wasm_test` is temporary disabled for my debugging of OOM killed in CI.

**Comment by lizan** [comment]
@PiotrSikora I fixed all build stuff, ptal.

File: test/extensions/bootstrap/wasm/test_data/BUILD [hunk 14]
```diff
@@ -140,3 +135,7 @@ wasm_cc_binary(
         "@proxy_wasm_cpp_sdk//:proxy_wasm_intrinsics",
     ],
 )
+
+# TODO(Shikugawa): Use bazel to build this with rewritten Rust test data from
+# https://github.com/istio/envoy/pull/251
+exports_files(["logging_rust.wasm"])
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/BUILD:141]: This is already available, see: https://github.com/envoyproxy/envoy-wasm/pull/611

But if we want to get this in first, and leave the checked in `.wasm` file, then we should probably leave the sources as well, otherwise there is no way to regenerate it.

File: test/extensions/bootstrap/wasm/test_data/Makefile [hunk 15]
```diff
@@ -1,3 +0,0 @@
-all: logging_rust.wasm
-
-include ../../../../../api/wasm/rust/Makefile.base
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/Makefile:3]: This should stay until we build `logging_rust.wasm` with Bazel.

File: test/extensions/bootstrap/wasm/BUILD [hunk 16]
```diff
@@ -20,7 +20,7 @@ envoy_extension_cc_test(
         "//test/extensions/bootstrap/wasm/test_data:bad_signature_cpp.wasm",
         "//test/extensions/bootstrap/wasm/test_data:emscripten_cpp.wasm",
         "//test/extensions/bootstrap/wasm/test_data:logging_cpp.wasm",
-        "//test/extensions/bootstrap/wasm/test_data:modules",
+        "//test/extensions/bootstrap/wasm/test_data:logging_rust.wasm",
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/BUILD:23]: The same should be done in `test/extensions/filters/{http,network}/wasm/BUILD`.

File: test/extensions/filters/http/wasm/test_data/body_rust/Cargo.toml [hunk 17]
```diff
@@ -1,18 +0,0 @@
...
-proxy-wasm = "0.1"
-log = "0.4"
-
-[lib]
-crate-type = ["cdylib"]
-
-[profile.release]
-lto = true
-opt-level = 3
-panic = "abort"
```
Comment (PiotrSikora) [inline test/extensions/filters/http/wasm/test_data/body_rust/Cargo.toml:18]: Why is this removed?

File: test/extensions/bootstrap/wasm/test_data/logging_rust.rs [hunk 18]
```diff
@@ -0,0 +1,49 @@
+use log::{debug, error, info, trace, warn};
```
Comment (PiotrSikora) [inline test/extensions/bootstrap/wasm/test_data/logging_rust.rs:1]: This file is copied (not moved) from `test/extensions/bootstrap/wasm/test_data/logging_rust/src/lib.rs`, and it's currently unused.

**Comment by lizan** [comment]
@PiotrSikora I reverted those unnecessary changes
